File: HackerRank/Dictionaries_and_Hashmaps/Java/ransomNote.java
import java.util.Hashtable;

public class ransomNote {

    public static void main(String[] args){
        /* setup for challenge */
        System.out.println("== Ransom Note (Java) ==");

        String m = "Hello World";
        String n = "Hello world";

        String[] magazine = m.split(" ");
        String[] note = n.split(" ");

        checkMagazine(magazine, note);
    }

    // Complete the checkMagazine function below.
    static void checkMagazine(String[] magazine, String[] note) {
        //since case is important, I don't need to do any conversions

        // first, create a hashtable for all the words in the magazine    
        Hashtable<String, Integer> magWordCount = new Hashtable<>();

        for(String word : magazine){
            if(magWordCount.get(word) == null){
                magWordCount.put(word, 1);// add the word to the hashtable if it's not there
            }
            else{
                magWordCount.put(word, magWordCount.get(word) + 1);// if the word's already there, add one to the count
            }
        }

        // now go through the note and see if it can be written from the magazine words
        
        int count = 0;

        for(String word : note){
            if(magWordCount.get(word) == null || magWordCount.get(word) == 0){
                // null is for if the word was never in the list. 0 is for if it was removed
                System.out.println("No");
                break;
            }
            else{
                magWordCount.put(word, magWordCount.get(word) - 1);
                count++;
            }
        }

        if(count == note.length){
            System.out.println("Yes");
        }
    }
}

File: HackerRank/Dictionaries_and_Hashmaps/Python/ransomNote.py
# Complete the checkMagazine function below.
def checkMagazine(magazine, note):
    ''' given two lists (m and n), I want to find out if
    every element in n is contained within m.
    Print 'Yes' if it is, and 'No' if not.
    Case sensitivity means I can leave the text as is
    '''

    # convert magazine into a dictionary
    magaDict = {}

    for i in range(len(magazine)):
        if magazine[i] in magaDict.keys():
            magaDict[magazine[i]] += 1
        else:
            magaDict[magazine[i]] = 1

    # now check if magazine contains every element in note
    count = 0
    
    for i in range(len(note)):
        if note[i] in magaDict.keys():
            # the word shows up in magazine
            if magaDict[note[i]] < 1:
                # this means there weren't enough occurences of the word
                print('No')
                break
            else:
                magaDict[note[i]] -= 1
                count += 1
        else:
            # the word doesn't show up in magazine
            print('No')
            break

    if count == len(note):
        print('Yes')

def main():
    # setup for solution

    m = 'Hello World'
    n = 'Hello world'

    magazine = m.split()
    note = n.split()

    checkMagazine(magazine, note)

main()